SOD.Common/Helpers/PlayerStatus.cs

```diff
@@ -0,0 +1,181 @@
+        /// </summary>
+        /// <param name="key">The key identifying the modifier.</param>
+        public void ToggleIllegalStatusModifier(string key)
+        {
+            if (IllegalStatusModifierDictionary.ContainsKey(key))
+            {
+                RemoveIllegalStatusModifier(key);
+                return;
+            }
+            SetIllegalStatusModifier(key);
+        }
+
+        internal IEnumerator UpdateIllegalStatusModifiers()
+        {
+            var setAsIllegal = false;
+            var modifierKeysToRemove = new List<string>();
+            var deltaTime = 0.0f;
+            while (true)
+            {
+                yield return new WaitForEndOfFrame();
+
+                if (Lib.Time.IsGamePaused)
+                {
+                    continue;
+                }
+
+                setAsIllegal = false;
+                deltaTime = UnityEngine.Time.deltaTime;
+
+                foreach (var key in IllegalStatusModifierDictionary.Keys)
+                {
+                    var entry = IllegalStatusModifierDictionary[key];
+                    if (!entry.UseTime)
+                    {
+                        setAsIllegal = true;
+                        continue;
+                    }
+                    entry.TimeRemainingSec -= deltaTime;
+                    if (entry.TimeRemainingSec > 0.0f)
+                    {
+                        setAsIllegal = true;
+                        continue;
+                    }
+                    // Expired timer, so ignore it and queue the modifier to be
+                    // removed
+                    modifierKeysToRemove.Add(entry.Key);
+                }
+
+                modifierKeysToRemove.ForEach(key => IllegalStatusModifierDictionary.Remove(key));
+                modifierKeysToRemove.Clear();
+
+                if (Player.Instance.illegalStatus == setAsIllegal)
+                {
+                    // Avoid constant changing back and forth
+                    Player.Instance.illegalActionTimer = deltaTime * 2.0f;
+                    continue;
+                }
+
+                // We want our illegal status to fall off as soon as the
+                // dictionary is empty, but to last until at least the next
+                // frame.
+                // Use deltaTime * 2.0f to prevent issues where the timer
+                // immediately expires during the UpdateIllegalStatus call
+                var args = new IllegalActionStatusArgs(setAsIllegal);
+                Lib.PlayerStatus.OnSetIllegalActionStatus(args, false);
+                Player.Instance.illegalActionActive = setAsIllegal;
+                Player.Instance.illegalActionTimer = deltaTime * 2.0f;
+                Player.Instance.UpdateIllegalStatus();
+                Lib.PlayerStatus.OnSetIllegalActionStatus(args, true);
+            }
```

Comment:
I think it would be good to make the coroutine null at the end if the dictionary count == 0